### PR TITLE
HDDS-594. SCM CA: DN sends CSR and uses certificate issued by SCM.

### PR DESCRIPTION
**Pull request by ajayydv**

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 25 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 23 | Maven dependency ordering for branch |
| +1 | mvninstall | 1000 | trunk passed |
| +1 | compile | 71 | trunk passed |
| +1 | checkstyle | 26 | trunk passed |
| +1 | mvnsite | 71 | trunk passed |
| +1 | shadedclient | 717 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 118 | trunk passed |
| +1 | javadoc | 67 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 12 | Maven dependency ordering for patch |
| +1 | mvninstall | 74 | the patch passed |
| +1 | compile | 67 | the patch passed |
| +1 | javac | 67 | the patch passed |
| +1 | checkstyle | 24 | the patch passed |
| +1 | mvnsite | 65 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 745 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 131 | the patch passed |
| +1 | javadoc | 60 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 61 | common in the patch failed. |
| -1 | unit | 72 | container-service in the patch failed. |
| +1 | asflicense | 29 | The patch does not generate ASF License warnings. |
| | | 3518 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdds.security.x509.certificate.client.TestDefaultCertificateClient |
|   | hadoop.ozone.container.common.TestDatanodeStateMachine |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-547/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/547 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  |
| uname | Linux ccb71df5b025 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 9e53088 |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/1/artifact/out/patch-unit-hadoop-hdds_common.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/1/artifact/out/patch-unit-hadoop-hdds_container-service.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/1/testReport/ |
| Max. process+thread count | 445 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/common hadoop-hdds/container-service U: hadoop-hdds |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/1/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 0 | Docker mode activated. |
| -1 | patch | 7 | https://github.com/apache/hadoop/pull/547 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help. |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/547 |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/2/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 0 | Docker mode activated. |
| -1 | patch | 6 | https://github.com/apache/hadoop/pull/547 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help. |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/547 |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/3/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 24 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 20 | Maven dependency ordering for branch |
| +1 | mvninstall | 980 | trunk passed |
| +1 | compile | 72 | trunk passed |
| +1 | checkstyle | 29 | trunk passed |
| +1 | mvnsite | 75 | trunk passed |
| +1 | shadedclient | 716 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 115 | trunk passed |
| +1 | javadoc | 58 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 10 | Maven dependency ordering for patch |
| +1 | mvninstall | 76 | the patch passed |
| +1 | compile | 67 | the patch passed |
| +1 | javac | 67 | the patch passed |
| -0 | checkstyle | 22 | hadoop-hdds: The patch generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0) |
| +1 | mvnsite | 61 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 741 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 128 | the patch passed |
| +1 | javadoc | 62 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 62 | common in the patch failed. |
| -1 | unit | 71 | container-service in the patch failed. |
| +1 | asflicense | 25 | The patch does not generate ASF License warnings. |
| | | 3464 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdds.security.x509.certificates.TestCertificateSignRequest |
|   | hadoop.hdds.security.x509.certificate.client.TestDefaultCertificateClient |
|   | hadoop.ozone.container.common.TestDatanodeStateMachine |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-547/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/547 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  |
| uname | Linux a6fe03809c56 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 9fcd89a |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/5/artifact/out/diff-checkstyle-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/5/artifact/out/patch-unit-hadoop-hdds_common.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/5/artifact/out/patch-unit-hadoop-hdds_container-service.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/5/testReport/ |
| Max. process+thread count | 423 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/common hadoop-hdds/container-service U: hadoop-hdds |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/5/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 35 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 32 | Maven dependency ordering for branch |
| +1 | mvninstall | 1066 | trunk passed |
| +1 | compile | 88 | trunk passed |
| +1 | checkstyle | 31 | trunk passed |
| +1 | mvnsite | 83 | trunk passed |
| +1 | shadedclient | 827 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 144 | trunk passed |
| +1 | javadoc | 97 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 10 | Maven dependency ordering for patch |
| +1 | mvninstall | 80 | the patch passed |
| +1 | compile | 72 | the patch passed |
| +1 | javac | 72 | the patch passed |
| +1 | checkstyle | 25 | the patch passed |
| +1 | mvnsite | 77 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 829 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 149 | the patch passed |
| +1 | javadoc | 67 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 86 | common in the patch failed. |
| +1 | unit | 68 | container-service in the patch passed. |
| +1 | asflicense | 29 | The patch does not generate ASF License warnings. |
| | | 3922 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdds.security.x509.certificate.client.TestDefaultCertificateClient |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-547/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/547 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  |
| uname | Linux 5c5da289cb7d 3.13.0-153-generic #203-Ubuntu SMP Thu Jun 14 08:52:28 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 9fcd89a |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/4/artifact/out/patch-unit-hadoop-hdds_common.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/4/testReport/ |
| Max. process+thread count | 313 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/common hadoop-hdds/container-service U: hadoop-hdds |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/4/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by xiaoyuyao** [comment]
@ajayydv , thanks for the update. Please fix the unit test failure testGenerateCSRWithInvalidParams and checkstyle issue. I will post another comments on the 2nd revision shortly.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 28 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 34 | Maven dependency ordering for branch |
| +1 | mvninstall | 1254 | trunk passed |
| +1 | compile | 79 | trunk passed |
| +1 | checkstyle | 30 | trunk passed |
| -1 | mvnsite | 28 | container-service in trunk failed. |
| +1 | shadedclient | 833 | branch has no errors when building and testing our client artifacts. |
| -1 | findbugs | 24 | container-service in trunk failed. |
| +1 | javadoc | 65 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 13 | Maven dependency ordering for patch |
| +1 | mvninstall | 79 | the patch passed |
| +1 | compile | 68 | the patch passed |
| +1 | javac | 68 | the patch passed |
| +1 | checkstyle | 26 | the patch passed |
| +1 | mvnsite | 66 | the patch passed |
| -1 | whitespace | 1 | The patch has 75 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply |
| -1 | whitespace | 1 | The patch 19849  line(s) with tabs. |
| +1 | shadedclient | 1016 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 130 | the patch passed |
| +1 | javadoc | 36 | common in the patch passed. |
| +1 | javadoc | 26 | hadoop-hdds_container-service generated 0 new + 7 unchanged - 1 fixed = 7 total (was 8) |
||| _ Other Tests _ |
| -1 | unit | 76 | common in the patch failed. |
| +1 | unit | 60 | container-service in the patch passed. |
| +1 | asflicense | 29 | The patch does not generate ASF License warnings. |
| | | 4197 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdds.security.x509.certificate.client.TestDefaultCertificateClient |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-547/6/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/547 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  |
| uname | Linux bc20fcc97965 4.4.0-139-generic #165~14.04.1-Ubuntu SMP Wed Oct 31 10:55:11 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 5fcea54 |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| mvnsite | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/6/artifact/out/branch-mvnsite-hadoop-hdds_container-service.txt |
| findbugs | v3.1.0-RC1 |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/6/artifact/out/branch-findbugs-hadoop-hdds_container-service.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/6/artifact/out/whitespace-eol.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/6/artifact/out/whitespace-tabs.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/6/artifact/out/patch-unit-hadoop-hdds_common.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/6/testReport/ |
| Max. process+thread count | 340 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/common hadoop-hdds/container-service U: hadoop-hdds |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/6/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 0 | Docker mode activated. |
| -1 | patch | 8 | https://github.com/apache/hadoop/pull/547 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help. |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/547 |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/7/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 0 | Docker mode activated. |
| -1 | patch | 9 | https://github.com/apache/hadoop/pull/547 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help. |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/547 |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/8/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/TestHddsSecureDatanodeInit.java [hunk 1]
```diff
@@ -252,23 +251,17 @@ public static void callQuietly(Callable closure) {
...
     keyCodec.writePrivateKey(privateKey);
-    VersionResponse versionResponse = VersionResponse
-        .newBuilder()
-        .setVersion(0)
-        .addValue(OzoneConsts.SCM_ID, "123")
-        .addValue(OzoneConsts.CLUSTER_ID, "123")
-        .build();
     PKCS10CertificationRequest csr =
-        service.getCSR(client, conf, versionResponse);
+        service.getCSR(client, conf);
```
[the remark this answers is not in the document]
Comment (ajayydv) [reply hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/TestHddsSecureDatanodeInit.java]: This is unit test for only CSR generation on DN side. Updated Default Approver on SCM side to add scm id and cluster id if not already present. Added corresponding tests in TestDefaultCAServer.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/HddsDatanodeService.java [hunk 2]
```diff
@@ -252,38 +250,37 @@ public void initializeCertificateClient(OzoneConfiguration config)
...
    * Get SCM signed certificate and store it using certificate client.
+   * @param client
+   * @param config
    * */
   private void getSCMSignedCert(CertificateClient client,
       OzoneConfiguration config) throws IOException {
+    PKCS10CertificationRequest csr = getCSR(client, config);
+    SCMSecurityProtocol secureScmClient =
+        HddsUtils.getScmSecurityClient(config,
+            HddsUtils.getScmAddressForSecurityProtocol(config));
```
[the remark this answers is not in the document]
Comment (ajayydv) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/HddsDatanodeService.java]: done.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/HddsDatanodeService.java [hunk 3]
```diff
@@ -192,6 +218,92 @@ public void start(Object service) {
...
+      break;
+    case GETCERT:
+      getSCMSignedCert(dnCertClient, config);
+      LOG.info("Successfully stored SCM signed certificate.");
+      break;
+    case FAILURE:
+      LOG.error("DN security initialization failed.");
+      throw new RuntimeException("DN security initialization failed.");
+    case RECOVER:
+      LOG.error("DN security initialization failed. OM certificate is " +
```
[the remark this answers is not in the document]
Comment (ajayydv) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/HddsDatanodeService.java]: Added case info to all cases.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/HddsDatanodeService.java [hunk 4]
```diff
@@ -252,38 +250,37 @@ public void initializeCertificateClient(OzoneConfiguration config)
...
-      String pemEncodedCert = secureScmClient.getDataNodeCertificate(
-          datanodeDetails.getProtoBufMessage(), getEncodedString(csr));
-
-      try {
-        X509Certificate x509Certificate =
-            CertificateCodec.getX509Certificate(pemEncodedCert);
-        client.storeCertificate(x509Certificate);
-      } catch (IOException | CertificateException e) {
-        LOG.error("Error while storing SCM signed certificate.", e);
-        throw new RuntimeException(e);
```
[the remark this answers is not in the document]
Comment (ajayydv) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/HddsDatanodeService.java]: I think IPC layer already has logic to retry for socket timeout related errors. If it fails there (unrecoverable), then its best to shutdown DN.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/HddsDatanodeService.java [hunk 5]
```diff
@@ -295,12 +292,9 @@ public PKCS10CertificationRequest getCSR(CertificateClient client,
     builder.setCA(false)
         .setKey(keyPair)
         .setConfiguration(config)
-        .setScmID(version.getValue(SCM_ID))
```
[the remark this answers is not in the document]
Comment (ajayydv) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/HddsDatanodeService.java]: Latest commit handles that.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/HddsDatanodeService.java [hunk 6]
```diff
@@ -192,6 +218,92 @@ public void start(Object service) {
...
+    LOG.info("Initializing secure Datanode.");
+
+    CertificateClient.InitResponse response = dnCertClient.init();
+    LOG.info("Init response: {}", response);
+    switch (response) {
+    case SUCCESS:
+      LOG.info("Initialization successful.");
+      break;
+    case GETCERT:
+      getSCMSignedCert(dnCertClient, config);
```
[the remark this answers is not in the document]
Comment (ajayydv) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/HddsDatanodeService.java]: done.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/HddsDatanodeService.java [hunk 7]
```diff
@@ -192,6 +218,92 @@ public void start(Object service) {
...
+    }
+
+
+  }
+
+  /**
+   * Creates CSR for DN.
+   * */
+  @VisibleForTesting
+  public PKCS10CertificationRequest getCSR(CertificateClient client,
```
[the remark this answers is not in the document]
Comment (ajayydv) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/HddsDatanodeService.java]: done

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsUtils.java [hunk 8]
```diff
@@ -161,6 +173,27 @@ public static InetSocketAddress getScmAddressForBlockClients(
         .orElse(ScmConfigKeys.OZONE_SCM_BLOCK_CLIENT_PORT_DEFAULT));
   }
 
+  /**
+   * Create a scm security client.
+   *
+   * @return {@link ScmBlockLocationProtocol}
```
[the remark this answers is not in the document]
Comment (ajayydv) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsUtils.java]: done.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 0 | Docker mode activated. |
| -1 | patch | 6 | https://github.com/apache/hadoop/pull/547 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help. |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/547 |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/9/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by ajayydv** [comment]
rebased with trunk.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 0 | Docker mode activated. |
| -1 | patch | 20 | https://github.com/apache/hadoop/pull/547 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help. |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/547 |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/10/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by xiaoyuyao** [comment]
why some of the changes from HDDS-1118 are shown up as part of the files changed?

**Comment by ajayydv** [comment]
@xiaoyuyao rebase seems to have pushed spurious commit. Forced reset it to intended commits.

**Comment by xiaoyuyao** [comment]
+1 pending Jenkins.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 33 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 4 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 12 | Maven dependency ordering for branch |
| +1 | mvninstall | 993 | trunk passed |
| +1 | compile | 73 | trunk passed |
| +1 | checkstyle | 31 | trunk passed |
| +1 | mvnsite | 79 | trunk passed |
| +1 | shadedclient | 723 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 122 | trunk passed |
| +1 | javadoc | 67 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 10 | Maven dependency ordering for patch |
| +1 | mvninstall | 78 | the patch passed |
| +1 | compile | 70 | the patch passed |
| +1 | javac | 70 | the patch passed |
| -0 | checkstyle | 26 | hadoop-hdds: The patch generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0) |
| +1 | mvnsite | 68 | the patch passed |
| -1 | whitespace | 0 | The patch has 75 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply |
| -1 | whitespace | 1 | The patch 19849  line(s) with tabs. |
| +1 | shadedclient | 893 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 128 | the patch passed |
| +1 | javadoc | 63 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 61 | common in the patch failed. |
| -1 | unit | 52 | container-service in the patch failed. |
| +1 | asflicense | 32 | The patch does not generate ASF License warnings. |
| | | 3684 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdds.security.x509.certificate.client.TestDefaultCertificateClient |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-547/12/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/547 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  |
| uname | Linux 932fd68b209e 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 45f976f |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/12/artifact/out/diff-checkstyle-hadoop-hdds.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/12/artifact/out/whitespace-eol.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/12/artifact/out/whitespace-tabs.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/12/artifact/out/patch-unit-hadoop-hdds_common.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/12/artifact/out/patch-unit-hadoop-hdds_container-service.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/12/testReport/ |
| Max. process+thread count | 410 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/common hadoop-hdds/container-service U: hadoop-hdds |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/12/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 30 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 4 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 22 | Maven dependency ordering for branch |
| +1 | mvninstall | 1301 | trunk passed |
| +1 | compile | 79 | trunk passed |
| +1 | checkstyle | 30 | trunk passed |
| +1 | mvnsite | 72 | trunk passed |
| +1 | shadedclient | 785 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 124 | trunk passed |
| +1 | javadoc | 64 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 11 | Maven dependency ordering for patch |
| +1 | mvninstall | 78 | the patch passed |
| +1 | compile | 69 | the patch passed |
| +1 | javac | 69 | the patch passed |
| +1 | checkstyle | 25 | the patch passed |
| +1 | mvnsite | 65 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 770 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 139 | the patch passed |
| +1 | javadoc | 60 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 73 | common in the patch failed. |
| -1 | unit | 80 | container-service in the patch failed. |
| +1 | asflicense | 25 | The patch does not generate ASF License warnings. |
| | | 3940 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdds.security.x509.certificate.client.TestDefaultCertificateClient |
|   | hadoop.ozone.container.common.TestDatanodeStateMachine |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-547/13/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/547 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  |
| uname | Linux 1f4f9c25055d 4.4.0-138-generic #164~14.04.1-Ubuntu SMP Fri Oct 5 08:56:16 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 09a9938 |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/13/artifact/out/patch-unit-hadoop-hdds_common.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/13/artifact/out/patch-unit-hadoop-hdds_container-service.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/13/testReport/ |
| Max. process+thread count | 399 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/common hadoop-hdds/container-service U: hadoop-hdds |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-547/13/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by ajayydv** [comment]
unit test failure seem unrelated, pass locally.

**Comment by xiaoyuyao** [comment]
I've tested the patch with docker and validated the DN initialization works as expected. The ci author/acceptance test does not seem to be related to this patch. Will merge it shortly.

**Pull request merged**